File: iosApp/iosApp/Modules/Degen/Cult/CultProposal/CultProposalViewModel.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import UIKit

struct CultProposalViewModel {

    let title: String
    let proposals: [ProposalDetails]
    let selectedIndex: Int

    struct ProposalDetails {
        
        let name: String
        let status: Status
        let guardianInfo: GuardianInfo?
        let summary: Summary
        let documentsInfo: DocumentsInfo
        let tokenomics: Tokenomics
        
        struct GuardianInfo {
            
            let title: String
            let name: String
            let nameValue: String
            let socialHandle: String
            let socialHandleValue: String
            let wallet: String
            let walletValue: String
        }
        
        struct Status {
            
            let title: String
            let color: UIColor // HEX
        }
        
        struct Summary {
            
            let title: String
            let summary: String
        }
        
        struct DocumentsInfo {
            
            let title: String
            let documents: [Document]
            
            struct Document {
                
                let title: String
                let items: [Item]
                
                enum Item {
                    
                    case link(displayName: String, url: URL)
                    case note(String)
                }
            }
        }
        
        struct Tokenomics {
            
            let title: String
            let rewardAllocation: String
            let rewardAllocationValue: String
            let rewardDistribution: String
            let rewardDistributionValue: String
            let projectETHWallet: String
            let projectETHWalletValue: String
        }
    }
    
    var titleIcon: Data {
        
        UIImage(named: "degen-cult-icon")!.pngData()!
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/degenCultProposal/CultProposalViewModel.kt
package com.sonsofcrypto.web3walletcore.modules.degenCultProposal

data class CultProposalViewModel(
    val title: String,
    val proposals: List<ProposalDetails>,
    val selectedIndex : Int,
){
    data class ProposalDetails(
        val name: String,
        val status: Status,
        val guardianInfo: GuardianInfo?,
        val summary: Summary,
        val documentsInfo: DocumentsInfo,
        val tokenomics : Tokenomics,
    ) {
        data class GuardianInfo(
            val title: String,
            val name: String,
            val nameValue: String,
            val socialHandle: String,
            val socialHandleValue: String,
            val wallet: String,
            val walletValue: String,
        )

        enum class Status { PENDING, CLOSED }

        data class Summary(
            val title: String,
            val summary: String,
        )

        data class DocumentsInfo(
            val title: String,
            val documents: List<Document>,
        ) {
            data class Document(
                val title: String,
                val items: List<Item>,
            ) {
                sealed class Item {
                    data class Link(val displayName: String, val url: String): Item()
                    data class Note(val text: String): Item()
                }
            }
        }

        data class Tokenomics(
            val title: String,
            val rewardAllocation: String,
            val rewardAllocationValue: String,
            val rewardDistribution: String,
            val rewardDistributionValue: String,
            val projectETHWallet: String,
            val projectETHWalletValue: String,
        )
    }
}